internal/cache/cache.go

```diff
@@ -4,111 +4,138 @@ import (
 	"PersonalAccountAPI/internal/models"
 	"PersonalAccountAPI/internal/usecase"
 	"context"
-	"mime/multipart"
 	"sync"
+	"time"
 
 	"github.com/pkg/errors"
 )
 
 type CacheDecorator struct {
-	mx           sync.RWMutex
-	userProvider *usecase.UserUsecase
-	userMap      map[int]models.UserRequest
-	userLoginMap map[string]models.UserRequest
-	// второстепенно ttl time.Duration
+	userProvider usecase.UserProvider
+	mu           sync.RWMutex
+	userMap      map[int]models.WrapUser
+	ttl          time.Duration
 }
 
-func New(user *usecase.UserUsecase) *CacheDecorator {
-	return &CacheDecorator{
-		mx:           sync.RWMutex{},
+func New(user usecase.UserProvider, ttl time.Duration) *CacheDecorator {
+	cache := &CacheDecorator{
 		userProvider: user,
-		userMap:      make(map[int]models.UserRequest),
-		userLoginMap: make(map[string]models.UserRequest),
+		mu:           sync.RWMutex{},
+		userMap:      make(map[int]models.WrapUser),
+		ttl:          ttl,
+	}
+
+	return cache
+}
+
+func (c *CacheDecorator) delete() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for key, wrapped := range c.userMap {
+		if time.Now().After(wrapped.TTL) {
+			delete(c.userMap, key)
+		}
+	}
+}
 
-		// ttl
+func (c *CacheDecorator) RunCleaner(ctx context.Context, checkInterval time.Duration) {
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.delete()
+		//WARNING: not work
+		case <-ctx.Done():
+			return
+		}
 	}
 }
 
 func (c *CacheDecorator) getUserMapValue(key int) (models.UserRequest, bool) {
-	c.mx.RLock()
-	defer c.mx.RUnlock()
+	c.mu.RLock()
+	defer c.mu.RUnlock()
 
-	user, exists := c.userMap[key]
-	return user, exists
+	wrapped := c.userMap[key]
+	if time.Now().After(wrapped.TTL) {
+		return models.UserRequest{}, false
+	}
+
+	wrapped, exists := c.userMap[key]
+	return wrapped.User, exists
 }
```

Comment:
Сломанная логика -> если объекта нет в мапе, значит во wrapped.TTL лежит zeroValue.
Старайся делать как можно меньше походов в мапу/базу.
Суть метода вернуть объект из мапы, сначала проверяем наличие и дальше протухание.

```suggestion
func (c *CacheDecorator) getUserMapValue(key int) (models.UserRequest, bool) {
    c.mu.RLock()
    defer c.mu.RUnlock()

    if wrapped, exists := c.userMap[key]; exists && time.Now().Before(wrapped.TTL) { //один запрос к мапе
        return wrapped.User, true
    }

    return models.UserRequest{}, false
}
```

Reply:
Не соглашусь, вроде в go как и в cpp есть short-circuiting

Reply:
done